### PR TITLE
[ new ] Add monomorphisation and unification to `elab-util-extra`

### PR DESCRIPTION
**Pull request by UARTman**
This PR upstreams my work on monomorphisation (done in https://github.com/DepTyCheck/dtc-utils).

It is not yet *fully* ready for upstreaming. Ideally, I'd like to do the following before it becomes ready to merge:
- [x] Get the final approval on the monomorphiser interface
- [x] Clean up the code further
- [x] Document all modules and functions
- [x] Either catch or document in code a few edge cases, such as failing cast declarations if monomorphisation task contains unused arguments
- [x] Write a minimal suite of golden-tests for both the unifier and the monomorphiser

### CODE REVIEW

File: elab-util-extra/tests/specialise-data/_shared/Shared.idr [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    (Just _, Just _) => verifyDecEqs polyTy specTy specVals
+    (Nothing, Just _) => fail "Specialised type implements DecEq, while polymorhpic type doesn't"
+    (Just _, Nothing) => fail "Polymorphic type implements DecEq, while specialised type doesn't"
+    (Nothing, Nothing) => pure ()
+  case (!(search $ Show polyTy), !(search $ Show specTy)) of
+    (Just _, Just _) => verifyShows polyTy specTy specVals
+    (Nothing, Just _) => fail "Specialised type implements Show, while polymorhpic type doesn't"
+    (Just _, Nothing) => fail "Polymorphic type implements Show, while specialised type doesn't"
+    (Nothing, Nothing) => pure ()
+  case (!(search $ Eq polyTy), !(search $ Show specTy)) of
```
Comment (buzden) [inline elab-util-extra/tests/specialise-data/_shared/Shared.idr]: Inconsistent, should be `Eq` on both sides.

I recommend merging all these checks in something like
```idris
for_ [`(DecEq), `(Show), `(Eq)] $ \expr => do case (!(search $ ~expr polyTy), ...) of ...
```

File: elab-util-extra/tests/specialise-data/specdata-009-finConst/Test.idr [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+e1' = %runElab verifySpecialisation (Fin 1) Fin1
+  [ `( FZ )
+  ]
+
+%runElab specialiseData' (Fin 2) "Fin2"
+
+--- Workaround for https://github.com/idris-lang/Idris2/issues/3651
+e2' = %runElab verifySpecialisation (Fin 2) Fin2
+  [ `( FZ )
+  , `( FS FZ ) -- TODO: Implement Num auto-derivation?
```
Comment (buzden) [inline elab-util-extra/tests/specialise-data/specdata-009-finConst/Test.idr]: not a `Num auto-derivation`, but `non-Num fromInteger auto-derivation`

File: elab-util-extra/tests/unify-with-compiler/_shared/Shared.idr [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+public export
+showUnify' : {l,r : Nat} -> Vect l Arg -> TTImp -> Vect r Arg -> TTImp -> Elab ()
+showUnify' lhsAs lhs rhsAs rhs = withUnify' lhsAs lhs rhsAs rhs $ logMsg "" 0 . show
+
+public export
+assertFV : UnificationResult -> Name -> TTImp -> Elab ()
+assertFV ur n t = do
+  let Just res = lookup n ur.fullResult
+  | _ => fail "Free variable \{show n} doesn't have a value"
+  if res == t
```
Comment (buzden) [inline elab-util-extra/tests/unify-with-compiler/_shared/Shared.idr]: You can use `when` from `base`

File: elab-util-extra/src/Deriving/SpecialiseData.idr [hunk 4]
```diff
@@ -0,0 +1,906 @@
...
+import Data.DPair
+import Data.List1
+import Data.Vect
+import Data.List
+import Data.List.Quantifiers
+import Data.Either
+import Data.SortedMap
+import Data.SortedSet
+import Data.SortedMap.Dependent
+import public Language.Mk -- TODO: Fix mk so that it works without this hack
```
[the remark this answers is not in the document]
Comment (UARTman) [reply elab-util-extra/src/Deriving/SpecialiseData.idr]: In the end, there was no real issue. It just turns out that using getMk in an elaborator script still requires the caller to have Mk imported.

File: elab-util-extra/src/Language/Reflection/Expr.idr [hunk 5]
```diff
@@ -55,6 +55,17 @@ cleanupNamedHoles = mapTTImp $ \case
   IHole {} => implicitFalse
   e        => e
 
+||| Run `cleanupNamedHoles` over a `PiInfo`'s inner `TTImp`
+public export
+cleanupPiInfo : PiInfo TTImp -> PiInfo TTImp
+cleanupPiInfo (DefImplicit t) = DefImplicit $ cleanupNamedHoles t
+cleanupPiInfo p               = p
```
Comment (buzden) [inline elab-util-extra/src/Language/Reflection/Expr.idr]: looks like `cleanupPiInfo = map cleanupNamedHoles`

File: elab-util-extra/src/Deriving/SpecialiseData.idr [hunk 6]
```diff
@@ -0,0 +1,1100 @@
...
+TaskLambda b => TaskLambda ({_ : a} -> b)
+
+export
+TaskLambda b => TaskLambda ((0 _ : a) -> b)
+
+export
+TaskLambda b => TaskLambda ((0 _ : a) => b)
+
+export
+TaskLambda b => TaskLambda ({0 _ : a} -> b)
```
Comment (buzden) [inline elab-util-extra/src/Deriving/SpecialiseData.idr]: Looks like this interface is used only in `specialisedData*`-functions, so I'd suggest to move it near to them

File: elab-util-extra/src/Language/Reflection/Unify/Interface.idr [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+%name DependencyGraph dg, depGraph
+
+%runElab derive "DependencyGraph" [Show]
+
+export
+Eq DependencyGraph where
+  (==) (MkDG a b c d e f) (MkDG a' b' c' d' e' f') with (decEq a a')
+   (==) (MkDG a b c d e f) (MkDG a' b' c' d' e' f') | Yes p =
+    a == a' && b == (rewrite p in b') && c == (rewrite p in c') &&
+      d == (rewrite p in d') && e == (rewrite p in e') && f == (rewrite p in f')
```
Comment (buzden) [inline elab-util-extra/src/Language/Reflection/Unify/Interface.idr]: You should match `Yes p` on `Yes Refl`, replace `a'` to `a` in the pattern-matching LHS, and thus get rid of all `rewrite`s

**Review by buzden — APPROVED** [review]
